File: src/kvs.erl
-module(kvs).

-behaviour(application).

-behaviour(supervisor).

-include_lib("stdlib/include/assert.hrl").

-include("api.hrl").

-include("metainfo.hrl").

-include("stream.hrl").

-include("cursors.hrl").

-include("kvs.hrl").

-include("backend.hrl").

-export([dump/0,
         db/0,
         metainfo/0,
         ensure/1,
         ensure/2,
         seq_gen/0,
         keys/1,
         fields/1,
         defined/2,
         field/2,
         setfield/3,
         remove/1]).

-export([join/2, seq/3]).

-export(?API)

.

-export(?STREAM)

.

-export([init/1, start/2, stop/1]).

-record('$msg', {id, next, prev, user, msg}).

init([]) -> {ok, {{one_for_one, 5, 10}, []}}.

start(_, _) ->
    supervisor:start_link({local, kvs}, kvs, []).

stop(_) -> ok.

dba() -> application:get_env(kvs, dba, kvs_mnesia).

seq_dba() -> application:get_env(kvs, dba_seq, kvs_mnesia).

db()  -> (dba()):db().

kvs_stream() ->
    application:get_env(kvs, dba_st, kvs_stream).

% kvs api

all(Table) -> all(Table, #kvs{mod = dba(), db = db()}).

delete(Table, Key) ->
    delete(Table, Key, #kvs{mod = dba(), db = db()}).

get(Table, Key) ->
    (?MODULE):get(Table, Key, #kvs{mod = dba(), db = db()}).

index(Table, K, V) ->
    index(Table, K, V, #kvs{mod = dba()}).

keys(Feed) ->
    keys(Feed, #kvs{mod = dba(), db = db()}).

key_match(Feed, Id) ->
  key_match(Feed, Id, #kvs{mod = dba(), db=db()}).

match(Record) ->
    match(Record, #kvs{mod = dba()}).

index_match(Record, Index) ->
    index_match(Record, Index, #kvs{mod = dba()}).

join() -> join([], #kvs{mod = dba(), db = db()}).

dump() -> dump(#kvs{mod = dba()}).

join(Node) -> join(Node, #kvs{mod = dba(), db = db()}).

leave() -> leave(#kvs{mod = dba(), db = db()}).

destroy() -> destroy(#kvs{mod = dba(), db = db()}).

count(Table) -> count(Table, #kvs{mod = dba()}).

put(Record) -> (?MODULE):put(Record, #kvs{mod = dba(), db = db()}).

stop() -> stop_kvs(#kvs{mod = dba()}).

start() -> start(#kvs{mod = dba()}).

ver() -> ver(#kvs{mod = dba()}).

dir() -> dir(#kvs{mod = dba()}).

feed(Key) ->
    feed(Key, #kvs{mod = dba(), st = kvs_stream(), db = db()}).

seq([], DX)    -> seq([], DX, #kvs{mod = kvs_rocks});
seq(Table, DX) -> seq(Table, DX, #kvs{mod = seq_dba()}).

remove(Rec, Feed) ->
    remove(Rec, Feed, #kvs{mod = dba(), st = kvs_stream(), db = db()}).

put(Records, #kvs{mod = Mod, db = Db}) when is_list(Records) ->
    Mod:put(Records, Db);
put(Record, #kvs{mod = Mod, db = Db}) -> Mod:put(Record, Db).

get(RecordName, Key, #kvs{mod = Mod, db = Db}) ->
    Mod:get(RecordName, Key, Db).

delete(Tab, Key, #kvs{mod = Mod, db = Db}) ->
    Mod:delete(Tab, Key, Db).

delete_range(Feed, Last, #kvs{mod=DBA, db=Db}) ->
    DBA:delete_range(Feed,Last,Db).

count(Tab, #kvs{mod = DBA}) -> DBA:count(Tab).

index(Tab, Key, Value, #kvs{mod = DBA}) ->
    DBA:index(Tab, Key, Value).

keys(Feed, #kvs{mod = DBA, db = Db}) ->
    DBA:keys(Feed, Db).

key_match(Feed, Id, #kvs{mod = DBA, db = Db}) ->
    DBA:key_match(Feed, Id, Db).

match(Record, #kvs{mod = DBA}) ->
    DBA:match(Record).

index_match(Record, Index, #kvs{mod = DBA}) ->
    DBA:index_match(Record, Index).

seq(Tab, Incr, #kvs{mod = DBA}) -> DBA:seq(Tab, Incr).

dump(#kvs{mod = Mod}) -> Mod:dump().

feed(Tab, #kvs{st = Mod, db = Db}) -> Mod:feed(Tab, Db).

remove(Rec, Feed, #kvs{st = Mod, db = Db}) ->
    Mod:remove(Rec, Feed, Db).

all(Tab, #kvs{mod = DBA, db = Db}) -> DBA:all(Tab, Db).

start(#kvs{mod = DBA}) -> DBA:start().

stop_kvs(#kvs{mod = DBA}) -> DBA:stop().

join(Node, #kvs{mod = DBA, db = Db}) -> DBA:join(Node, Db).

leave(#kvs{mod = DBA, db = Db}) -> DBA:leave(Db).

destroy(#kvs{mod = DBA, db = Db}) -> DBA:destroy(Db).

ver(#kvs{mod = DBA}) -> DBA:version().

dir(#kvs{mod = DBA}) -> DBA:dir().

% stream api

top(X) -> (kvs_stream()):top(X).

top(X,#kvs{db = Db}) -> (kvs_stream()):top(X,Db).

bot(X) -> (kvs_stream()):bot(X).

bot(X,#kvs{db = Db})         -> (kvs_stream()):bot(X,Db).

next(X)                      -> (kvs_stream()):next(X).

next(X,#kvs{db = Db})        -> (kvs_stream()):next(X,Db).

prev(X)                      -> (kvs_stream()):prev(X).

prev(X,#kvs{db = Db})        -> (kvs_stream()):prev(X,Db).

drop(X)                      -> (kvs_stream()):drop(X).

drop(X,#kvs{db = Db})        -> (kvs_stream()):drop(X,Db).

take(X)                      -> (kvs_stream()):take(X).

take(X,#kvs{db = Db})        -> (kvs_stream()):take(X,Db).

save(X)                      -> (kvs_stream()):save(X).

save(X,#kvs{db = Db})        -> (kvs_stream()):save(X,Db).

remove(X)                    -> (kvs_stream()):remove(X).

cut(X)                       -> (kvs_stream()):cut(X).

cut(X,#kvs{db = Db})         -> (kvs_stream()):cut(X, Db).

add(X)                       -> (kvs_stream()):add(X).

add(X,#kvs{db = Db})         -> (kvs_stream()):add(X,Db).

append(X, Y)                 -> (kvs_stream()):append(X, Y).

append(X, Y, #kvs{db = Db})  -> (kvs_stream()):append(X, Y, Db).

load_reader(X)               -> (kvs_stream()):load_reader(X).

load_reader(X,#kvs{db = Db}) -> (kvs_stream()):load_reader(X,Db).

writer(X)                    -> (kvs_stream()):writer(X).

writer(X,#kvs{db = Db})      -> (kvs_stream()):writer(X,Db).

reader(X)                    -> (kvs_stream()):reader(X).

reader(X,#kvs{db = Db})      -> (kvs_stream()):reader(X,Db).

% unrevisited

ensure(#writer{} = X) -> ensure(X,#kvs{mod = dba(), db = db()}).
ensure(#writer{id = Id},#kvs{} = X) ->
    case kvs:get(writer, Id, X) of
        {error, _} ->
            kvs:save(kvs:writer(Id, X)),
            ok;
        {ok, _} -> ok
    end.

cursors() ->
    lists:flatten([[{T#table.name, T#table.fields}
                    || #table{name = Name} = T
                           <- (M:metainfo())#schema.tables,
                       Name == reader orelse Name == writer]
                   || M <- modules()]).

% metainfo api

tables() ->
    lists:flatten([(M:metainfo())#schema.tables
                   || M <- modules()]).

modules() -> application:get_env(kvs, schema, []).

metainfo() ->
    #schema{name = kvs, tables = core() ++ test_tabs()}.

core() ->
    [#table{name = id_seq,
            fields = record_info(fields, id_seq), keys = [thing]}].

test_tabs() ->
    [#table{name = '$msg',
            fields = record_info(fields, '$msg')}].

table(Name) when is_atom(Name) ->
    lists:keyfind(Name, #table.name, tables());
table(_) -> false.

seq_gen() ->
    Init = fun (Key) ->
                   case kvs:get(id_seq, Key) of
                       {error, _} ->
                           {Key, kvs:put(#id_seq{thing = Key, id = 0})};
                       {ok, _} -> {Key, skip}
                   end
           end,
    [Init(atom_to_list(Name))
     || {Name, _Fields} <- cursors()].

initialize(Backend, Module) ->
    [begin
         Backend:create_table(T#table.name,
                              [{attributes, T#table.fields},
                               {T#table.copy_type, [node()]},
                               {type, T#table.type}]),
         [Backend:add_table_index(T#table.name, Key)
          || Key <- T#table.keys],
         T
     end
     || T <- (Module:metainfo())#schema.tables].

fields(Table) when is_atom(Table) ->
    case table(Table) of
        false -> [];
        T -> T#table.fields
    end.

defined(TableRecord, Field) ->
    FieldsList = fields(element(1, TableRecord)),
    lists:member(Field, FieldsList).

field(TableRecord, Field) ->
    FieldsList = fields(element(1, TableRecord)),
    Index = string:str(FieldsList, [Field]) + 1,
    element(Index, TableRecord).

setfield(TableRecord, Field, Value) ->
    FieldsList = fields(element(1, TableRecord)),
    Index = string:str(FieldsList, [Field]) + 1,
    setelement(Index, TableRecord, Value).

File: lib/KVS.ex
defmodule KVS do
  require Record
  @moduledoc "KVS Abstract Chain Database"
  Record.defrecord(:schema, name: nil, tables: [])
  Record.defrecord(:table, name: nil, container: false, type: :set, fields: [], keys: [], copy_type: :disc_copies, instance: {}, mappings: [])
  Record.defrecord(:writer, id: [], count: 0, args: [], cache: [], first: [])
  Record.defrecord(:reader, id: [], pos: 0, cache: [], args: [], feed: [], seek: [], count: 0, dir: 0)
  Record.defrecord(:id_seq, thing: [], id: 0)
  Record.defrecord(:it, id: [])
  Record.defrecord(:ite, id: [], next: [])
  Record.defrecord(:iter, id: [], next: [], prev: [])
  Record.defrecord(:kvs, mod: :kvs_mnesia, st: :kvs_stream, db: [], cx: [])
end

defmodule :kvs do
  use Application
  require KVS

  def start(_type, _args) do
    Supervisor.start_link([], strategy: :one_for_one, name: :kvs_sup)
  end

  def dba, do: Application.get_env(:kvs, :dba, :kvs_mnesia)
  def seq_dba, do: Application.get_env(:kvs, :dba_seq, :kvs_mnesia)
  def db, do: dba().db()
  def kvs_stream, do: Application.get_env(:kvs, :dba_st, :kvs_stream)

  # kvs api
  def all(feed) when is_binary(feed) or is_list(feed), do: feed(feed)
  def all(table), do: all(table, KVS.kvs(mod: dba(), db: db()))
  def delete(table, key), do: delete(table, key, KVS.kvs(mod: dba(), db: db()))
  def get(table, key), do: __MODULE__.get(table, key, KVS.kvs(mod: dba(), db: db()))
  def index(table, k, v), do: index(table, k, v, KVS.kvs(mod: dba()))
  def keys(feed), do: keys(feed, KVS.kvs(mod: dba(), db: db()))
  def key_match(feed, id), do: key_match(feed, id, KVS.kvs(mod: dba(), db: db()))
  def match(record), do: match(record, KVS.kvs(mod: dba()))
  def index_match(record, index), do: index_match(record, index, KVS.kvs(mod: dba()))
  def join(), do: join([], KVS.kvs(mod: dba(), db: db()))
  def dump(), do: dump(KVS.kvs(mod: dba()))
  def join(node), do: join(node, KVS.kvs(mod: dba(), db: db()))
  def leave(), do: leave(KVS.kvs(mod: dba(), db: db()))
  def destroy(), do: destroy(KVS.kvs(mod: dba(), db: db()))
  def count(table), do: count(table, KVS.kvs(mod: dba()))
  def put(record), do: __MODULE__.put(record, KVS.kvs(mod: dba(), db: db()))
  def stop(), do: stop(KVS.kvs(mod: dba()))
  def start(), do: start(KVS.kvs(mod: dba()))
  def ver(), do: ver(KVS.kvs(mod: dba()))
  def dir(), do: dir(KVS.kvs(mod: dba()))
  def feed(key), do: feed(key, KVS.kvs(mod: dba(), st: kvs_stream(), db: db()))
  def seq([], dx), do: seq([], dx, KVS.kvs(mod: :kvs_rocks))
  def seq(table, dx), do: seq(table, dx, KVS.kvs(mod: seq_dba()))
  def remove(rec, feed), do: remove(rec, feed, KVS.kvs(mod: dba(), st: kvs_stream(), db: db()))

  def put(records, KVS.kvs(mod: mod, db: db)) when is_list(records), do: mod.put(records, db)
  def put(record, KVS.kvs(mod: mod, db: db)), do: mod.put(record, db)

  def get(record_name, key, KVS.kvs(mod: mod, db: db)), do: mod.get(record_name, key, db)
  def delete(tab, key, KVS.kvs(mod: mod, db: db)), do: mod.delete(tab, key, db)
  def delete_range(feed, last, KVS.kvs(mod: dba, db: db)), do: dba.delete_range(feed, last, db)
  def count(tab, KVS.kvs(mod: dba)), do: dba.count(tab)
  def index(tab, key, value, KVS.kvs(mod: dba)), do: dba.index(tab, key, value)
  def keys(feed, KVS.kvs(mod: dba, db: db)), do: dba.keys(feed, db)
  def key_match(feed, id, KVS.kvs(mod: dba, db: db)), do: dba.key_match(feed, id, db)
  def match(record, KVS.kvs(mod: dba)), do: dba.match(record)
  def index_match(record, index, KVS.kvs(mod: dba)), do: dba.index_match(record, index)
  def seq(tab, incr, KVS.kvs(mod: dba)), do: dba.seq(tab, incr)
  def dump(KVS.kvs(mod: mod)), do: mod.dump()
  def feed(tab, KVS.kvs(st: mod, db: db)), do: mod.feed(tab, db)
  def remove(rec, feed, KVS.kvs(st: mod, db: db)), do: mod.remove(rec, feed, db)
  def all(tab, KVS.kvs(mod: dba, db: db)), do: dba.all(tab, db)
  def start(KVS.kvs(mod: dba)), do: dba.start()
  def stop(KVS.kvs(mod: dba)), do: dba.stop()
  def join(node, KVS.kvs(mod: dba, db: db)), do: dba.join(node, db)
  def leave(KVS.kvs(mod: dba, db: db)), do: dba.leave(db)
  def destroy(KVS.kvs(mod: dba, db: db)), do: dba.destroy(db)
  def ver(KVS.kvs(mod: dba)), do: dba.version()
  def dir(KVS.kvs(mod: dba)), do: dba.dir()

  # stream api
  def top(x), do: kvs_stream().top(x)
  def top(x, KVS.kvs(db: db)), do: kvs_stream().top(x, db)
  def bot(x), do: kvs_stream().bot(x)
  def bot(x, KVS.kvs(db: db)), do: kvs_stream().bot(x, db)
  def next(x), do: kvs_stream().next(x)
  def next(x, KVS.kvs(db: db)), do: kvs_stream().next(x, db)
  def prev(x), do: kvs_stream().prev(x)
  def prev(x, KVS.kvs(db: db)), do: kvs_stream().prev(x, db)
  def drop(x), do: kvs_stream().drop(x)
  def drop(x, KVS.kvs(db: db)), do: kvs_stream().drop(x, db)
  def take(x), do: kvs_stream().take(x)
  def take(x, KVS.kvs(db: db)), do: kvs_stream().take(x, db)
  def save(x), do: kvs_stream().save(x)
  def save(x, KVS.kvs(db: db)), do: kvs_stream().save(x, db)
  def remove(x), do: kvs_stream().remove(x)
  def cut(x), do: kvs_stream().cut(x)
  def cut(x, KVS.kvs(db: db)), do: kvs_stream().cut(x, db)
  def add(x), do: kvs_stream().add(x)
  def add(x, KVS.kvs(db: db)), do: kvs_stream().add(x, db)
  def append(x, y), do: kvs_stream().append(x, y)
  def append(x, y, KVS.kvs(db: db)), do: kvs_stream().append(x, y, db)
  def load_reader(x), do: kvs_stream().load_reader(x)
  def load_reader(x, KVS.kvs(db: db)), do: kvs_stream().load_reader(x, db)
  def writer(x), do: kvs_stream().get_writer(x)
  def writer(x, KVS.kvs(db: db)), do: kvs_stream().get_writer(x, db)
  def reader(x), do: kvs_stream().get_reader(x)
  def reader(x, KVS.kvs(db: db)), do: kvs_stream().get_reader(x, db)

  def ensure(w = KVS.writer()), do: ensure(w, KVS.kvs(mod: dba(), db: db()))
  def ensure(KVS.writer(id: id), x = KVS.kvs()) do
    case get(:writer, id, x) do
      {:error, _} ->
        save(KVS.writer(id: id), x)
        :ok
      {:ok, _} ->
        :ok
    end
  end

  def cursors() do
    modules()
    |> Enum.flat_map(fn m ->
      m.metainfo()
      |> KVS.schema(:tables)
      |> Enum.filter(fn KVS.table(name: name) -> name == :reader or name == :writer end)
      |> Enum.map(fn KVS.table(name: n, fields: f) -> {n, f} end)
    end)
  end

  # metainfo api
  def tables() do
    modules() |> Enum.flat_map(fn m -> m.metainfo() |> KVS.schema(:tables) end)
  end

  def modules(), do: Application.get_env(:kvs, :schema, [])

  def metainfo() do
    KVS.schema(name: :kvs, tables: core() ++ test_tabs())
  end

  def core() do
    [
      KVS.table(name: :id_seq, fields: [:thing, :id], keys: [:thing]),
      KVS.table(name: :reader, fields: [:id, :pos, :cache, :args, :feed, :seek, :count, :dir]),
      KVS.table(name: :writer, fields: [:id, :count, :args, :cache, :first])
    ]
  end

  def test_tabs() do
    [KVS.table(name: :msg, fields: [:id, :next, :prev, :user, :msg])]
  end

  def get_table(name) when is_atom(name) do
    case Enum.find(tables(), fn KVS.table(name: n) -> n == name end) do
      nil -> false
      t -> t
    end
  end
  def get_table(_), do: false

  def seq_gen() do
    f = fn key ->
      case get(:id_seq, key) do
        {:error, _} -> {key, put(KVS.id_seq(thing: key, id: 0))}
        {:ok, _} -> {key, :skip}
      end
    end
    cursors() |> Enum.map(fn {name, _} -> f.(to_string(name)) end)
  end

  def initialize(backend, module) do
    module.metainfo()
    |> KVS.schema(:tables)
    |> Enum.map(fn t = KVS.table(name: name, fields: fields, copy_type: copy, type: type, keys: keys) ->
      backend.create_table(name, [attributes: fields, "#{copy}": [node()], type: type] ++ if(type == :set, do: [], else: [type: type]))
      Enum.each(keys, fn key -> backend.add_table_index(name, key) end)
      t
    end)
  end

  def fields(table_name) when is_atom(table_name) do
    case get_table(table_name) do
      false -> []
      t -> KVS.table(t, :fields)
    end
  end

  def defined(record, field) do
    field in fields(elem(record, 0))
  end

  def field(record, field) do
    fields_list = fields(elem(record, 0))
    index = Enum.find_index(fields_list, &(&1 == field)) + 1
    elem(record, index)
  end

  def setfield(record, field, value) do
    fields_list = fields(elem(record, 0))
    index = Enum.find_index(fields_list, &(&1 == field)) + 1
    put_elem(record, index, value)
  end
end
